src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -1396,35 +1399,65 @@ private async Task WriteStreamPropertiesAsync(SelectExpandNode selectExpandNode,
         {
             ISet<IEdmNavigationProperty> navigationProperties = selectExpandNode.SelectedNavigationProperties;
 
-            if (navigationProperties != null)
+            if (navigationProperties == null)
             {
-                IEnumerable<string> changedProperties = null;
-                if (null != resourceContext.EdmObject && resourceContext.EdmObject is IDelta changedObject)
+                yield break;
+            }
+
+            IEnumerable<string> changedProperties = null;
+
+            if (resourceContext.EdmObject != null && resourceContext.EdmObject is IDelta changedObject)
+            {
+                changedProperties = changedObject.GetChangedPropertyNames();
+
+                foreach (IEdmNavigationProperty navigationProperty in navigationProperties)
+                {
+                    if (changedProperties == null || changedProperties.Contains(navigationProperty.Name))
+                    {
+                        yield return new KeyValuePair<IEdmNavigationProperty, Type>(navigationProperty, typeof(IEdmChangedObject));
+                    }
+                }
+            }
+            else if (resourceContext.ResourceInstance != null && resourceContext.ResourceInstance is IDelta deltaObject)
```

Comment:
```suggestion
            else if (resourceContext.ResourceInstance is IDelta deltaObject)
```
Since `resourceContext.ResourceInstance is IDelta` would evaluate to `false` if `resourceContext.ResourceInstance` is `null` - I BELIEVE!